### PR TITLE
#66 Implement ElasticSearch accessing

### PR DESCRIPTION
**Pull request by ilisau**
## PR-Codex overview
This PR adds Elasticsearch support to the project, allowing for more efficient searching of tours. It also introduces new domain fields and updates existing ones to support this functionality.

### Detailed summary
- Added Elasticsearch dependency to pom.xml
- Created `ElasticsearchClient` and `RestClient` beans in `WebConfig.java`
- Added `Field` interface and implemented new fields such as `Country`, `TourType`, `CateringType`, `ArrivalTime`, `DepartureTime`, `Price`, and `DayDuration`
- Updated `TourCriteria` to include a list of `Field` objects
- Updated `TourService` to use `CriteriaQuery` and `Field` objects for more efficient searching
- Updated `TourController` to include a new endpoint for searching tours by description
- Updated `TourCriteriaMapper` to map `TourCriteriaDto` to a list of `Field` objects
- Updated `HotelDto` and `AddressDto` to include Elasticsearch annotations for indexing fields

> The following files were skipped due to too many changes: `src/main/java/com/solvd/qaprotours/web/dto/AddressDto.java`, `src/main/java/com/solvd/qaprotours/web/dto/TourCriteriaDto.java`, `src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java`, `src/test/java/com/solvd/qaprotours/config/TestConfig.java`, `src/test/java/com/solvd/qaprotours/service/impl/TourServiceTests.java`, `src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java`, `src/main/java/com/solvd/qaprotours/web/dto/TourDto.java`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java [hunk 1]
```diff
@@ -12,12 +16,30 @@
...
 import reactor.core.publisher.Mono;
 
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
 /**
  * @author Varya Petrova, Lisov Ilya
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java:44]: @ilisau each author should be placed from new line

File: src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java [hunk 2]
```diff
@@ -40,13 +64,162 @@ public Flux<Tour> getAll(final Pagination pagination,
...
+        return Flux.just(tours.toArray(new Tour[0]));
+    }
+
+    @SneakyThrows
+    private List<Tour> getAllByCriteria(final TourCriteria tourCriteria,
+                                        final int currentPage,
+                                        final int pageSize) {
+        List<QueryBuilder> queryBuilders = new ArrayList<>();
+        CriteriaQuery searchQuery = new CriteriaQuery(new Criteria());
+
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java:85]: remove empty lines

File: src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java [hunk 3]
```diff
@@ -40,13 +64,162 @@ public Flux<Tour> getAll(final Pagination pagination,
...
+        List<String> countries = tourCriteria.getCountries();
+        if (countries != null && !countries.isEmpty()) {
+            Criteria criteria = new Criteria("country");
+            criteria.in(countries);
+            searchQuery.addCriteria(criteria);
+        }
+
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java:92]: the same here

File: src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java [hunk 4]
```diff
@@ -40,13 +64,162 @@ public Flux<Tour> getAll(final Pagination pagination,
...
+            return Flux.just(tours.toArray(new Tour[0]));
+        }
+        List<Tour> tours = getAllByCriteria(tourCriteria,
+                pagination.getCurrentPage(),
+                pagination.getPageSize());
+        return Flux.just(tours.toArray(new Tour[0]));
+    }
+
+    @SneakyThrows
+    private List<Tour> getAllByCriteria(final TourCriteria tourCriteria,
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/service/impl/TourServiceImpl.java:80]: this method is an absolute monster, unmaintainable
simply if you want to add new fields to criteria - you will need to change this method
make it through objects

each field can be applied to criteria through `apply()` method
like this: 

```java
interface Field {
void apply();

class Country implements Field {
}
etc...
```

so, each field will be applied in a declarative way
now, if you want to add/remove fields in criteria you will add/remove field object into object chain

File: src/main/java/com/solvd/qaprotours/service/impl/fake/FakeTourService.java [hunk 5]
```diff
@@ -27,6 +27,14 @@ public Flux<Tour> getAll(final Pagination pagination,
         return Flux.concat(tour1, tour2);
     }
 
+    @Override
+    public Flux<Tour> getAll(final Pagination pagination,
+                             final String description) {
+        Mono<Tour> tour1 = getById(1L);
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/service/impl/fake/FakeTourService.java:33]: give better names for your variables

**Review by h1alexbel — CHANGES_REQUESTED** [review]

**Comment by ilisau** [comment]
So I need to do smth like this or maybe store all criteria in another form?

File: src/main/java/com/solvd/qaprotours/domain/field/ArrivalTimeField.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package com.solvd.qaprotours.domain.field;
+
+import lombok.Data;
+import org.springframework.data.elasticsearch.core.query.Criteria;
+import org.springframework.data.elasticsearch.core.query.CriteriaQuery;
+
+import java.time.LocalDateTime;
+
+@Data
+public class ArrivalTimeField implements Field {
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/field/ArrivalTimeField.java:10]: @ilisau you can simplify class name
just remove `Field` suffix

File: src/main/java/com/solvd/qaprotours/domain/field/CountryField.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package com.solvd.qaprotours.domain.field;
+
+import lombok.Data;
+import org.springframework.data.elasticsearch.core.query.Criteria;
+import org.springframework.data.elasticsearch.core.query.CriteriaQuery;
+
+import java.util.List;
+
+@Data
+public class CountryField implements Field {
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/field/CountryField.java:10]: the same here

File: src/main/java/com/solvd/qaprotours/domain/field/DayDurationField.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package com.solvd.qaprotours.domain.field;
+
+import lombok.Data;
+import org.springframework.data.elasticsearch.core.query.Criteria;
+import org.springframework.data.elasticsearch.core.query.CriteriaQuery;
+
+@Data
+public class DayDurationField implements Field {
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/field/DayDurationField.java:8]: the same here

File: src/main/java/com/solvd/qaprotours/domain/field/TourTypeField.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.solvd.qaprotours.domain.tour.Tour;
+import lombok.Data;
+import org.springframework.data.elasticsearch.core.query.Criteria;
+import org.springframework.data.elasticsearch.core.query.CriteriaQuery;
+
+import java.util.List;
+
+@Data
+public class TourTypeField implements Field {
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/field/TourTypeField.java:11]: the same here

File: src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java [hunk 10]
```diff
@@ -1,29 +1,54 @@
...
-    private List<Tour.TourType> tourTypes;
-    private Integer stars;
-    private List<Tour.CateringType> cateringTypes;
-    private List<Integer> coastLines;
-    private LocalDateTime arrivedAt;
-    private LocalDateTime leavedAt;
-    private Integer dayDuration;
-    private BigDecimal minCost;
-    private BigDecimal maxCost;
+    private CountryField country;
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java:19]: @ilisau try `List<Field>`

File: src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java [hunk 11]
```diff
@@ -1,29 +1,54 @@
...
+    private ArrivalTimeField arrivalTime;
+    private DepartureTimeField departureTime;
+    private DayDurationField dayDuration;
+    private PriceField price;
+
+    /**
+     * Creates a TourCriteria object with empty criteria.
+     */
+    public TourCriteria() {
+        country = new CountryField();
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java:31]: @ilisau looks so unflexible

use List of fields

File: src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java [hunk 12]
```diff
@@ -1,29 +1,54 @@
...
+        dayDuration = new DayDurationField();
+        price = new PriceField();
+    }
+
+    /**
+     * Applies all criteria to fields.
+     * @param searchQuery CriteriaQuery object
+     */
+    public void apply(final CriteriaQuery searchQuery) {
+        this.country.apply(searchQuery);
```
Comment (h1alexbel) [inline src/main/java/com/solvd/qaprotours/domain/tour/TourCriteria.java:45]: @ilisau obvious code duplication, adding more fields your method will grow

**Review by h1alexbel — CHANGES_REQUESTED** [review]

**Comment by ilisau** [comment]
@h1alexbel I need to map fields somewhere, because, as I explored, Mapstruct can't do it by itself

**Review by h1alexbel — CHANGES_REQUESTED** [review]

**Review by h1alexbel — APPROVED** [review]

**Comment by h1alexbel** [comment]
@ilisau nice

**Pull request merged**